File: devcamp2018/app/src/main/java/com/psato/devcamp/presentation/MVVM/QueryViewModelArc.java
package com.psato.devcamp.presentation.MVVM;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;

import com.psato.devcamp.data.entity.ShowResponse;
import com.psato.devcamp.interactor.usecase.show.SearchShows;

import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.functions.Consumer;

public class QueryViewModelArc extends ViewModel {

    private MutableLiveData<String> result = new MutableLiveData<>();

    private MutableLiveData<String> query = new MutableLiveData<>();

    private MutableLiveData<Boolean> showLoading = new MutableLiveData<>();

    private MutableLiveData<Boolean> searchEnabled = new MutableLiveData<>();

    private Observer<String> queryObserver;

    private SearchShows searchShows;

    @Inject
    public QueryViewModelArc(SearchShows searchShows) {
        this.searchShows = searchShows;
        showLoading.setValue(false);
        searchEnabled.setValue(false);
        result.setValue("");
        query.setValue("");
        queryObserver = query -> searchEnabled.setValue(!TextUtils.isEmpty(query));
        query.observeForever(queryObserver);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        query.removeObserver(queryObserver);
    }

    public void onQueryClick(View view) {
        searchShow(query.getValue());
    }

    private void searchShow(String value) {
        if (searchShows != null) {
            showLoading.setValue(true);
            searchShows.unsubscribe();
            searchShows.setQuery(value);
            Date start = new Date();
            searchShows.execute((Consumer<List<ShowResponse>>) title -> {
                Date end = new Date();
                Log.e("SATO", "SATO - Time: " + (end.getTime() - start.getTime())/1000 + "s");
                showLoading.setValue(false);
                result.setValue(title.get(0).getName());
            }, throwable -> showLoading.setValue(false));
        }
    }

    public MutableLiveData<Boolean> getSearchEnabled() {
        return searchEnabled;
    }

    public MutableLiveData<Boolean> getShowLoading() {
        return showLoading;
    }

    public MutableLiveData<String> getQuery() {
        return query;
    }

    public MutableLiveData<String> getResult() {
        return result;
    }

}

File: devcamp2018/app/src/main/java/com/psato/devcamp/presentation/MVVM/QueryViewModelArc.kt
package com.psato.devcamp.presentation.MVVM

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModel
import android.text.TextUtils
import android.view.View
import com.psato.devcamp.interactor.usecase.show.SearchShows
import javax.inject.Inject

class QueryViewModelArc @Inject
constructor(private val searchShows: SearchShows?) : ViewModel() {

    val result = MutableLiveData<String>()

    val query = MutableLiveData<String>()

    val showLoading = MutableLiveData<Boolean>()

    val searchEnabled = MutableLiveData<Boolean>()

    private val queryObserver: Observer<String>

    init {
        showLoading.value = false
        searchEnabled.value = false
        result.value = ""
        query.value = ""
        queryObserver = Observer { query -> searchEnabled.value = !TextUtils.isEmpty(query) }
        query.observeForever(queryObserver)
    }

    override fun onCleared() {
        super.onCleared()
        query.removeObserver(queryObserver)
    }

    fun onQueryClick(view: View) {
        searchShow()
    }

    private fun searchShow() {
        if (searchShows != null) {
            showLoading.value = true
            searchShows.unsubscribe()
            searchShows.query = query.value
            searchShows.execute({ title: String ->
                showLoading.value = false
                result.value = title
            }, { throwable ->
                showLoading.value = false
                result.value = throwable.message
            })
        }
    }

}
